### DISCUSSION TITLE
Troubleshooting $search variable accessibility in Select Field modifyQueryUsing() method

### DISCUSSION THREAD

**Discussion by manojhl** [filamentphp/filament#13168] [Help]
### Package

Form builder

### Package Version

v3.2

### How can we help you?

According to the document, the $search variable should be accessible within 

https://filamentphp.com/docs/3.x/forms/fields/select#customizing-the-relationship-query

> ### Customizing the relationship query
> You may customize the database query that retrieves options using the third parameter of the `relationship()` method:
> 
> ```php
> use Filament\Forms\Components\Select;
> use Illuminate\Database\Eloquent\Builder;
>  
> Select::make('author_id')
>     ->relationship(
>         name: 'author',
>         titleAttribute: 'name',
>         modifyQueryUsing: fn (Builder $query) => $query->withTrashed(),
>     )
> ```
> 
> If you would like to access the current search query in the `modifyQueryUsing` function, you can inject `$search`.

I tried adding $search in the `modifyQueryUsing` but i got an error

```php
 Select::make('skills')
  ->multiple()
  ->relationship(
        titleAttribute: 'name',
        modifyQueryUsing: function (Builder $query, ?string $search) {
            return $query->withTrashed()
                ->where('name', 'LIKE', '%' . $search . '%')
                ->orderByRaw(
                    "CASE
                        WHEN name LIKE ? THEN 1
                        WHEN name LIKE ? THEN 2
                        ELSE 3
                    END, name ASC",
                    [$search . '%', '%' . $search . '%']
                )
                ->limit(20);
        }
    )
```

> An attempt was made to evaluate a closure for [Filament\Forms\Components\Select], but [$search] was unresolvable.

`$search` parameter should be accessible inside the `modifyQueryUsing()` method, right? am i missing something or is this a bug?

**Comment by alisalehi1380** [comment]
I think you should add `searchable()` too.

```php
Select::make('skills')
  ->multiple()
  ->searchable() // this line
  ->relationship(
        titleAttribute: 'name',
        modifyQueryUsing: function (Builder $query, ?string $search) {
            return $query->withTrashed()
                ->where('name', 'LIKE', '%' . $search . '%')
                ->orderByRaw(
                    "CASE
                        WHEN name LIKE ? THEN 1
                        WHEN name LIKE ? THEN 2
                        ELSE 3
                    END, name ASC",
                    [$search . '%', '%' . $search . '%']
                )
                ->limit(20);
        }
    )
```

**Reply by manojhl** [reply]
nope even with `searchable()` i don't get access to `$search` parameters.

But i did find an alternative approach to override the search query using `getSearchResultsUsing()` method which gets access to `$search`

```php
Select::make('skills')
      ->multiple()
      ->relationship(titleAttribute: 'name')
      ->getSearchResultsUsing(fn (string $search): array => Skill::where('name', 'like', "%{$search}%")
          ->orderByRaw(
              "CASE
                  WHEN name LIKE ? THEN 1
                  WHEN name LIKE ? THEN 2
                  ELSE 3
              END, name ASC",
              ["{$search}%", "%{$search}%"]
          )->limit(50)->pluck('name', 'id')->toArray())
```

**Comment by ossipov** [comment]
Now when you input text in `Select multiple() relation()` field it seems to be searching data with `whereLike()`
I want to `explode()` search prompt by spaces and search with `whereLike()` for each word separetley.

Got to this discussion searching for text from docs:
> If you would like to access the current search query in the modifyQueryUsing function, you can inject $search.

Couldn't find a way to inject `$search` in `modifyQueryUsing()`

Also solution with `getSearchResultsUsing()` have access to `$search` but I didn't find a way to have access to selected `Record`

Maybe I'm missing something. Could anyone please help me 🙏
Now as a hacky workaround I'm using `%` inside prompt but it only helps when word sequence is predictable.

```md
# example data:
- Happy beautiful world
- World is happy and beautiful
```
```md
# prompt: 
happy world

# expected:
- Happy beautiful world
- World is happy and beautiful

# result:
empty
```
```md
# prompt:
happy%world

# result:
- Happy beautiful world
```

**Reply by ossipov** [reply]
Looks like there is a way to get `Record` inside `getSearchResultsUsing()`

```php
getSearchResultsUsing(function ($search) use ($form) {
  // $search
  // $form->getRecord()
})
```

Will try to use this aproach.
